File: src/ros_rqt_plugin/ros_rqt_plugin.cpp
#include "ros_rqt_plugin/ros_rqt_plugin.hpp"
#include <pluginlib/class_list_macros.h>
#include <QStringList>


namespace ros_rqt_plugin
{
    two_button_plugin::two_button_plugin()
    :   rqt_gui_cpp::Plugin(),
        widget_(0)
    {
        // Constructor is called first before initPlugin function, needless to say.
        // give QObjects reasonable names
        setObjectName("ros_rqt_plugin");;
    }

    void two_button_plugin::initPlugin(qt_gui_cpp::PluginContext & context)
    {
        // Access standalone command line arguments
        QStringList argv = context.argv();

        // Create QWidget
        widget_ = new QWidget();

        // Extend the widget with all attributes and children from UI file
        ui_.setupUi(widget_);

        // add widget to the user interface
        context.addWidget(widget_);

        // Define ROS publishers
        buttton_1_pub_ = getNodeHandle().advertise<std_msgs::Bool>("button_1_topic", 1);
        buttton_2_pub_ = getNodeHandle().advertise<std_msgs::Bool>("button_2_topic", 1);

        // Declare ROS msg_
        msg_.data = true;

        // Connect Qt Widgets
        connect(ui_.pushButton_1, SIGNAL(pressed()), this, SLOT(button_1_callback_()));
        connect(ui_.pushButton_2, SIGNAL(pressed()), this, SLOT(button_2_callback_()));
    }

    void two_button_plugin::shutdownPlugin()
    {
        ;
    }

    void two_button_plugin::saveSettings(
        qt_gui_cpp::Settings & plugin_settings,
        qt_gui_cpp::Settings & instance_settings
    ) const
    {
        ;
    }

    void two_button_plugin::restoreSettings(
        const qt_gui_cpp::Settings & plugin_settings,
        const qt_gui_cpp::Settings & instance_settings
    )
    {
        ;
    }

    void two_button_plugin::button_1_callback_()
    {
        // Publish msg_ as true
        buttton_1_pub_.publish(msg_);
    }

    void two_button_plugin::button_2_callback_()
    {
        // Publish msg_ as true
        buttton_2_pub_.publish(msg_);
    }

} // namespace ros_rqt_plugin

PLUGINLIB_EXPORT_CLASS(ros_rqt_plugin::two_button_plugin, rqt_gui_cpp::Plugin);

File: src/ros_rqt_plugin/ros_rqt_plugin.py
#!/usr/bin/env python

import os
import rospy
import rospkg

from qt_gui.plugin import Plugin
from python_qt_binding import loadUi
from python_qt_binding.QtWidgets import QWidget
# For add refresh infrom of button
# from python_qt_binding.QtGui import QIcon

from std_msgs.msg import Bool


class Buttons(Plugin):
    def __init__(self, context):
        super(Buttons, self).__init__(context)
        # Give QObjects reasonable names
        self.setObjectName('Buttons')

        # Process standalone plugin command-line arguments
        from argparse import ArgumentParser
        parser = ArgumentParser()
        # Add argument(s) to the parser.
        parser.add_argument("-q", "--quiet", action="store_true",
                            dest="quiet",
                            help="Put plugin in silent mode")
        args, unknowns = parser.parse_known_args(context.argv())
        if not args.quiet:
            print('arguments: ', args)
            print('unknowns: ', unknowns)

        # Create QWidget
        self._widget = QWidget()
        # Get path to UI file which should be in the "resource" folder of this package
        ui_file = os.path.join(rospkg.RosPack().get_path('ros_rqt_plugin'), 'resource', 'ros_rqt_plugin.ui')
        # Extend the widget with all attributes and children from UI file
        loadUi(ui_file, self._widget)
        # Give QObjects reasonable names
        self._widget.setObjectName('ros_rqt_plugin_ui')
        # Show _widget.windowTitle on left-top of each plugin (when
        # it's set in _widget). This is useful when you open multiple
        # plugins at once. Also if you open multiple instances of your
        # plugin at once, these lines add number to make it easy to
        # tell from pane to pane.
        if context.serial_number() > 1:
            self._widget.setWindowTitle(self._widget.windowTitle() +
                                        (' (%d)' % context.serial_number()))
        # Add widget to the user interface
        context.add_widget(self._widget)

        # Define ROS publisher
        self.button_1_pub = rospy.Publisher('button_1_topic', Bool, queue_size=1)
        self.button_2_pub = rospy.Publisher('button_2_topic', Bool, queue_size=1)

        # Define ROS msg
        self.msg = True

        # You can set button icon like so
        # For add refresh infrom of button
        # self._widget.pushButton_1.setIcon(QIcon.fromTheme('view-refresh'))

        # Connecting widget and ui
        self._widget.pushButton_1.pressed.connect(self.button_1_callback)
        self._widget.pushButton_2.pressed.connect(self.button_2_callback)

    def shutdown_plugin(self):
        # TODO unregister all publishers here
        self.button_1_pub.unregister()
        self.button_2_pub.unregister()

    def save_settings(self, plugin_settings, instance_settings):
        # TODO save intrinsic configuration, usually using:
        # instance_settings.set_value(k, v)
        pass

    def restore_settings(self, plugin_settings, instance_settings):
        # TODO restore intrinsic configuration, usually using:
        # v = instance_settings.value(k)
        pass

    # def trigger_configuration(self):
        # Comment in to signal that the plugin has a way to configure
        # This will enable a setting button (gear icon) in each dock widget
        # title bar
        # Usually used to open a modal configuration dialog

    def button_1_callback(self):
        self.button_1_pub.publish(self.msg)

    def button_2_callback(self):
        self.button_2_pub.publish(self.msg)